qcelemental/testing.py

```diff
@@ -0,0 +1,332 @@
+import sys
+import copy
+import math
+import pprint
+import collections
+
+import numpy as np
+
+
+class _TestComparisonError(Exception):
+    """Error when element or nuclide can't be identified."""
+
+    def __init__(self, msg):
+        self.message = '\nTestComparisonError: {}\n\n'.format(msg)
+
+
+def _success(label):
+    """Function to print a '*label*...PASSED' line to screen."""
+
+    print(f'\t{label:.<66}PASSED')
+    sys.stdout.flush()
+
+
+def compare_float(expected,
+                  computed,
+                  digits,
+                  label=None,
+                  *,
+                  rtol=1.e-16,
+                  equal_nan=False,
+                  passnone=False,
+                  return_message=False,
+                  verbose=1):
+    """Returns True if two floats or float arrays are element-wise equal within a tolerance.
+
+    Parameters
+    ----------
+    expected : float or float array-like
+        Reference value against which `computed` is compared.
+    computed : float or float array-like
+        Input value to compare against `expected`.
+    digits : int or float
+        Absolute tolerance (see formula below), expressed as decimal digits for comparison.
+        Values less than one are taken literally rather than as power.
+        So `1` means `atol=0.1` and `2` means `atol=0.01` but `0.04` means `atol=0.04`
+    label : str, optional
+        Label for passed and error messages. Defaults to calling function name.
+    rtol : float, optional
+        Relative tolerance (see formula below). By default set to zero so `digits` dominates.
+    equal_nan : bool, optional
+        Passed to np.isclose.
+    passnone : bool, optional
+        Return True when both expected and computed are None.
+    verbose : int, optional
+        Print passed message when >=1.
+
+    Returns
+    -------
+    allclose : bool
+        Returns True if `expected` and `computed` are equal within tolerance; False otherwise.
+    message : str, optional
+        When return_message=True, also return passed or error message.
+
+    Notes
+    -----
+    Akin to np.allclose.
+    Sets rtol to zero to match expected Psi4 behaviour, otherwise measured as:
+        absolute(computed - expected) <= (atol + rtol * absolute(expected))
+
+    """
+    label = label or sys._getframe().f_back.f_code.co_name
```

Comment:
I considered it. But when pytest passes, printing is hidden. When pytest fails, the auto-pass-message label shows which tests are failing and at which point in the test (for several asserts). If the goal is to use the fns as-is for pytest (rather than wrapping), I rather favor default at `verbose=1`. Does the printing disturb you in practice?

Reply:
I very often run `pytest -s`, could we have a global switch here to print or not based off a pytest flag?

Reply:
The other thing with printing that it would be interesting to try would be to use `sys._getframe().f_back.f_code.co_name` and have an extra `print()` if that is the first time seeing that name this run so that the printing comes out more canonically.

Reply:
> I very often run `pytest -s`, could we have a global switch here to print or not based off a pytest flag?

Yes, I see the trouble. Not seeing a way to access whether run with `-s`. What aspect of `pytest -s` are you using that `pytest` isn't providing?

Reply:
`pytest -s` is useful for ensuring that no extraneous prints are used, and also to allow debug information to be passed through when enabled.

Reply:
I guess on the flip side for your use case do you need to print out that each `assert` passed? Can we only print out information if the assert fails.

Reply:
historically (p4), the compare_ fn has printed success and at failure defers to the exception, which prints to screen and file.

I think who prints what where is very much up for renegotiation as we do the T/F None/raise inversion. I think on this first pass I want to get the comparison and info-present aspects right (`return_message=True` is setting things up for the Exception and the printing).

The case where I like printing for each passed `assert` is something like the below, where counting down to the failed one is handy.

```
    assert compare_values(expected, jrec['qcvars']['CURRENT ENERGY'].data, 7, tnm())
    assert compare_values(expected, jrec['qcvars']['DISPERSION CORRECTION ENERGY'].data, 7, tnm())
    assert compare_values(expected, jrec['qcvars']['3-BODY DISPERSION CORRECTION ENERGY'].data, 7, tnm())
    assert compare_values(expected, jrec['qcvars']['AXILROD-TELLER-MUTO 3-BODY DISPERSION CORRECTION ENERGY'].data, 7,
                          tnm())

    assert compare_arrays(gexpected, jrec['qcvars']['CURRENT GRADIENT'].data, 7, tnm())
    assert compare_arrays(gexpected, jrec['qcvars']['DISPERSION CORRECTION GRADIENT'].data, 7, tnm())
    assert compare_arrays(gexpected, jrec['qcvars']['3-BODY DISPERSION CORRECTION GRADIENT'].data, 7, tnm())
    assert compare_arrays(gexpected, jrec['qcvars']['AXILROD-TELLER-MUTO 3-BODY DISPERSION CORRECTION GRADIENT'].data,
                          7, tnm())
```

Reply:
Thinking about it, a better argument is perhaps that we do not want to deviate too far from canonical python testing. PyTest has its own nice custom raises for things like strings/sets/etc that derive from [this](https://pytest.readthedocs.io/en/2.9.1/assert.html#defining-your-own-assertion-comparison). I am not sure that this will work here as `conftest.py` will not propagate. 

A popular alternative is to have `assert_compare` which will raise an assertion error only if the internal return is False. This will hook nicely into pytest and pytest's tech will display the line number that the test failed at so it is an easy find.

Reply:
I rather like the passing printing, but I don't feel too strongly on this. so long as pass/fail and a nicely formatted string are accessible from running, I'll be content.

The pytest_assertrepr_compare are nifty. But for general utility I think I'd like to stick with plain python functions so not dependent on a testing framework or even being run in "test" as opposed to "production" mode. It could be that these fns are re-wrapped for various testing frameworks, even re-wrapped for QCA-flavored pytest. In that sense, silence and T/F returns probably best. Though I could be talked into T/AssertionError.

Reply:
`compare` = T/F + silence, `assert_compare` = T/AssertionError?

Reply:
where both in `testing.py` and the latter a light wrapper? sure.

Reply:
Great! I think that will work out.

Reply:
Another thing to consider is to print out to the logging module which you can turn on/off with pytest.